test/core/xds/xds_client_test.cc

```diff
@@ -3779,6 +3803,367 @@ TEST_F(XdsClientTest, AdsReadWaitsForHandleRelease) {
+}
+
+TEST_F(XdsClientTest, FallbackOnStartup) {
+  FakeXdsBootstrap::FakeXdsServer primary_server;
+  FakeXdsBootstrap::FakeXdsServer fallback_server("fallback_xds_server");
+  // Regular operation
+  InitXdsClient(FakeXdsBootstrap::Builder().SetServers(
+      {primary_server, fallback_server}));
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
+  auto primary_stream = WaitForAdsStream(primary_server);
+  ASSERT_NE(primary_stream, nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(primary_stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  TriggerConnectionFailure(primary_server,
+                           absl::UnavailableError("Primary server is down"));
+  EXPECT_THAT(
+      GetServerConnections(),
+      ::testing::ElementsAre(::testing::Pair(kDefaultXdsServerUrl, false),
+                             ::testing::Pair("fallback_xds_server", true)));
+  // XdsClient should have created an ADS stream.
+  auto fallback_stream = WaitForAdsStream(fallback_server);
+  ASSERT_NE(fallback_stream, nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  request = WaitForRequest(fallback_stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  // Send a response.
+  fallback_stream->SendMessageToClient(
+      ResponseBuilder(XdsFooResourceType::Get()->type_url())
+          .set_version_info("1")
+          .set_nonce("A")
+          .AddFooResource(XdsFooResource("foo1", 6))
+          .Serialize());
+  // XdsClient should have delivered the response to the watcher.
+  auto resource = watcher->WaitForNextResource();
+  ASSERT_NE(resource, nullptr);
+  EXPECT_EQ(resource->name, "foo1");
+  EXPECT_EQ(resource->value, 6);
+  request = WaitForRequest(fallback_stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"1", /*response_nonce=*/"A",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  // Recover to primary
+  primary_stream = WaitForAdsStream(primary_server);
+  primary_stream->SendMessageToClient(
+      ResponseBuilder(XdsFooResourceType::Get()->type_url())
+          .set_version_info("5")
+          .set_nonce("D")
+          .AddFooResource(XdsFooResource("foo1", 42))
+          .Serialize());
+  ASSERT_NE(primary_stream, nullptr);
+  fallback_stream.reset();
+  resource = watcher->WaitForNextResource();
```

Comment:
Please add a check here that `GetServerConnections()` shows a healthy channel to the primary server and no channel to the fallback server.